README.md
# deeprepair

## Environment
```
conda create -n myenv1 python=3.8
conda install pytorch torchvision tqdm scipy
```

```
conda create -n myenv2 python=2.7
conda install pytorch torchvision tqdm scipy scikit-learn matplotlib
```

## [CIFAR-10](https://github.com/yuchi1989/deeprepair/tree/master/exp_7/cifar10/confusion) prototype  


## [COCO](https://github.com/yuchi1989/deeprepair/tree/master/exp_9/coco/confusion_and_bias) prototype  


## Check Model Confusion
### cifar10
```
python3 repair_confusion_exp_newbn.py --net_type resnet --dataset cifar10 --depth 18 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --first 3 --second 5 --checkmodel
```

### cifar10 vgg
```
python3 repair_confusion_exp_newbn_vggbn.py --net_type resnet --dataset cifar10 --depth 18 --pretrained ../../../models/cifar10_vggbn_2_4/model_best.pth.tar --first 3 --second 5 --checkmodel
```

### cifar100
```
python3 repair_confusion_exp_newbn.py --net_type resnet --dataset cifar100 --depth 34 --pretrained ../../../models/cifar100_resnet34/model_best.pth.tar --first 35 --second 98 --checkmodel
```

### coco
```
python2 repair_confusion_bn.py --pretrained ../../../models/coco_original_model/model_best.pth.tar --ann_dir '../../../../coco/annotations' --image_dir '../../../../coco/' --first "person" --second "bus" --checkmodel
```

### coco gender
```
python2 repair_confusion_bn.py --pretrained ../../../models/cocogender_original_model/model_best.pth.tar --ann_dir '../../../../coco/annotations' --image_dir '../../../../coco/' --first "handbag" --second "woman" --checkmodel
```

### cifar10 2 pairs
```
python3 repair_confusion_exp_newbn.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --checkmodel --pair1a 3 --pair1b 5 --pair2a 1 --pair2b 9
```

## Check Model Bias
### cifar10
```
python3 repair_bias_exp_newbn.py --net_type resnet --dataset cifar10 --depth 18 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --first 5 --second 3 --third 2 --checkmodel
```

### cifar10 vgg
```
python3 repair_bias_exp_newbn_vggbn.py --net_type resnet --dataset cifar10 --depth 18 --pretrained ../../../models/cifar10_vggbn_2_4/model_best.pth.tar --first 5 --second 3 --third 2 --checkmodel
```

### cifar100
```
python3 repair_bias_exp_newbn.py --net_type resnet --dataset cifar100 --depth 34 --pretrained ../../../models/cifar100_resnet34/model_best.pth.tar --first 98 --second 35 --third 11 --checkmodel
```

### coco
```
python2 repair_bias_bn.py --pretrained ../../../models/coco_original_model/model_best.pth.tar --ann_dir '../../../../coco/annotations' --image_dir '../../../../coco/' --first "bus" --second "person" --third "clock" --checkmodel
```

### coco gender
```
python2 repair_bias_bn.py --pretrained ../../../models/cocogender_original_model/model_best.pth.tar --ann_dir '../../../../coco/annotations' --image_dir '../../../../coco/' --first "skis" --second "woman" --third "man" --checkmodel
```


## Get Instance
### [cifar10 get instance](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/cifar10_get_instance.py)

```
python3 cifar10_get_instance.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test  --cutmix_prob 0 --original ./runs/cifar10_resnet18_2_4/model_best.pth.tar --fix ./runs/cifar10_resnet_2_4_dogcat_dbr/model_best.pth.tar  --checkmodel --first 5 --second 3

python3 cifar10_get_instance.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test  --cutmix_prob 0 --original ./runs/cifar10_resnet18_2_4/model_best.pth.tar --fix ./runs/cifar10_resnet_2_4_dogcat_dbr/model_best.pth.tar  --checkmodel --first 3 --second 5

```

### [cifar100 get instance](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/cifar10_get_instance.py)

```
python3 cifar10_get_instance.py --net_type resnet --dataset cifar100 --depth 34 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test  --cutmix_prob 0 --original ./runs/cifar100_resnet34/model_best.pth.tar --fix ./runs/cifar100_resnet34_oversampling/model_best.pth.tar  --checkmodel --first 35 --second 98

python3 cifar10_get_instance.py --net_type resnet --dataset cifar100 --depth 34 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test  --cutmix_prob 0 --original ./runs/cifar100_resnet34/model_best.pth.tar --fix ./runs/cifar100_resnet34_oversampling/model_best.pth.tar  --checkmodel --first 98 --second 35

```

### [coco get instance](https://github.com/yuchi1989/deeprepair/blob/master/exp_9/coco/get_instance.py)

```
python2 get_instance.py --original original_model/model_best.pth.tar --fix fix_model/model_best.pth.tar --log_dir coco_confusion_repair_aug --first "person" --second "bus" --ann_dir '../../../../coco/annotations' --image_dir '../../../../coco/'
python2 get_instance.py --original original_model/model_best.pth.tar --fix fix_model/model_best.pth.tar --log_dir coco_confusion_repair_aug --first "bus" --second "person" --ann_dir '../../../../coco/annotations' --image_dir '../../../../coco/'
```

### [coco gender get instance bias](https://github.com/yuchi1989/deeprepair/blob/master/exp_9/coco_gender/get_instance_coco_gender_bias.py)

```
python2 get_instance_coco_gender_bias.py --original original_model/model_best.pth.tar --fix fix_model/model_best.pth.tar --log_dir coco_confusion_repair_aug --first "skis" --second "man" --third "woman" --ann_dir '../../../../coco/annotations' --image_dir '../../../../coco/'
```


## Fine-tuning Model
### cifar10
```
python3 repair_confusion_exp_oversampling.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_finetuned --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --first 3 --second 5 --weight 1
```
### cifar10 vgg
```
python3 repair_confusion_exp_oversampling_vggbn.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_vggbn_2_4_finetuned --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_vggbn_2_4/model_best.pth.tar --extra 128 --first 3 --second 5 --weight 1
```
### cifar100
```
python3 repair_confusion_exp_oversampling.py --net_type resnet --dataset cifar100 --depth 34 --batch_size 128 --lr 0.1 --expname cifar100_resnet34_finetuned --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar100_resnet34/model_best.pth.tar --extra 128 --first 35 --second 98 --weight 1
```
### coco
```
python2 repair_bias_exp_weighted_loss.py --pretrained ../../../models/coco_original_model/model_best.pth.tar --log_dir coco_original_model_finetuned --first "bus" --second "person" --third "clock" --ann_dir '../../../../coco/annotations' --image_dir '../../../../coco/' --weight 1 --class_num 80
```
### coco gender
```
python2 repair_confusion_exp_weighted_loss.py --pretrained ../../../models/cocogender_original_model/model_best.pth.tar --log_dir cocogender_original_model_finetuned --first "handbag" --second "woman" --ann_dir '../../../../coco/annotations' --image_dir '../../../../coco/' --weight 1 --class_num 81
```




## [CIFAR-10](https://github.com/yuchi1989/deeprepair/tree/master/exp_7/cifar10) experiments
### [CIFAR-10 confusion](https://github.com/yuchi1989/deeprepair/tree/master/exp_7/cifar10/confusion) experiments  

#### [orig](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/train_baseline.py):    

```
python3 train_baseline.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet18_2_4 --epochs 300 --beta 1.0 --cutmix_prob 0
```

#### [w-aug](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_exp_oversampling.py):  

```
python3 repair_confusion_exp_oversampling.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --first 3 --second 5 --weight 0.3
```

#### [w-bn](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_exp_newbn.py):  

```
python3 repair_confusion_exp_newbn.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --replace --first 3 --second 5 --ratio 0.2  
```

#### [w-os](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_exp_newbn_softmax.py):
```
python3 repair_confusion_exp_newbn_softmax.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --eta 0.3 --checkmodel --first 3 --second 5
```

#### [w-loss](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_exp_weighted_loss.py):
```
python3 repair_confusion_exp_weighted_loss.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --first 3 --second 5 --target_weight 0.4
```
#### [w-dbr](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_dbr.py):  

```
python3 repair_confusion_dbr.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 256 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --first 3 --second 5 --lam 0.7
```

### [CIFAR-10 bias](https://github.com/yuchi1989/deeprepair/tree/master/exp_7/cifar10/bias) experiments  

#### [w-aug](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/bias/repair_bias_exp_oversampling.py):  

```
python3 repair_bias_exp_oversampling.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --first 3 --second 5 --third 2 --weight 0.3
```

#### [w-bn](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/bias/repair_bias_exp_newbn.py):  

```
python3 repair_bias_exp_newbn.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --replace --ratio 0.2 --first 3 --second 5 --third 2
```

#### [w-os](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/bias/repair_bias_exp_newbn_softmax.py):
```
python3 repair_bias_exp_newbn_softmax.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --eta 0.3 --checkmodel --first 3 --second 5 --third 2
```

#### [w-loss](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/bias/repair_bias_exp_weighted_loss.py):
```
python3 repair_bias_exp_weighted_loss.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --first 3 --second 5 --third 2 --target_weight 0.4
```
#### [w-dbr](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/bias/repair_bias_dbr.py):  

```
python3 repair_bias_dbr.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 256 --lr 0.1 cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --first 3 --second 5 --third 2 --lam 0.7
```


### [CIFAR-10 2 pair confusion](https://github.com/yuchi1989/deeprepair/tree/master/exp_7/cifar10_multipair/confusion) experiments  


#### [w-aug](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/cifar10_multipair/repair_confusion_exp_oversampling.py):  

```
python3 repair_confusion_exp_oversampling.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128  --pair1a 3 --pair1b 5 --pair2a 1 --pair2b 9  --weight 0.3.0
```

#### [w-bn](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/cifar10_multipair/repair_confusion_exp_newbn.py):  

```
python3 repair_confusion_exp_newbn.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --replace  --pair1a 3 --pair1b 5 --pair2a 1 --pair2b 9   --ratio 0.4  
```

#### [w-os](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/cifar10_multipair/repair_confusion_exp_newbn_softmax.py):
```
python3 repair_confusion_exp_newbn_softmax.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128 --eta 0.3 --checkmodel  --pair1a 3 --pair1b 5 --pair2a 1 --pair2b 9  
```

#### [w-loss](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/cifar10_multipair/repair_confusion_exp_weighted_loss.py):
```
python3 repair_confusion_exp_weighted_loss.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar --extra 128  --pair1a 3 --pair1b 5 --pair2a 1 --pair2b 9   --target_weight 0.4
```
#### [w-dbr](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/cifar10_multipair/repair_confusion_dbr.py):  

```
python3 repair_confusion_dbr.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 256 --lr 0.1 --expname cifar10_resnet_2_4_dogcat_test --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_resnet18_2_4/model_best.pth.tar  --pair1a 3 --pair1b 5 --pair2a 1 --pair2b 9 --lam 0.7
```




### [CIFAR-10 VGG-BN confusion](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/train_baseline_vggbn.py) experiments
#### [orig](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/train_baseline.py):    

```
python3 train_baseline_vggbn.py  --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_vggbn_2_4 --epochs 300 --beta 1.0 --cutmix_prob 0
```

#### [w-aug](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_exp_oversampling_vggbn.py):  

```
python3 repair_confusion_exp_oversampling_vggbn.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_vggbn_oversampling --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_vggbn_2_4/model_best.pth.tar --extra 128 --first 3 --second 5 --weight 0.3
```

#### [w-bn](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_exp_newbn_vggbn.py):  

```
python3 repair_confusion_exp_newbn_vggbn.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_vggbn_newbn --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_vggbn_2_4/model_best.pth.tar --extra 128 --replace --first 3 --second 5 --ratio 0.4  
```

#### [w-os](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_exp_newbn_softmax_vggbn.py):
```
python3 repair_confusion_exp_newbn_softmax_vggbn.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_vggbn_softmax --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_vggbn_2_4/model_best.pth.tar --extra 128 --eta 0.1 --checkmodel --first 3 --second 5
```

#### [w-loss](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_exp_weighted_loss_vggbn.py):
```
python3 repair_confusion_exp_weighted_loss_vggbn.py --net_type resnet --dataset cifar10 --depth 18 --batch_size 128 --lr 0.1 --expname cifar10_vggbn_loss --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ../../../models/cifar10_vggbn_2_4/model_best.pth.tar --extra 128 --first 3 --second 5 --target_weight 0.4
```
#### [w-dbr](https://github.com/yuchi1989/deeprepair/blob/master/exp_7/cifar10/confusion/repair_confusion_dbr_vggbn.py):  

```

history.txt
 experiment1:

    Train baseline models   
        python3 train_baseline.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname DeepInspect_1 --epochs 300 --beta 1.0 --cutmix_prob 1
        # accuracy: 0.9464;
        python3 train_baseline.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname DeepInspect_0 --epochs 300 --beta 1.0 --cutmix_prob 0
        # accuracy: 0.933;


    Decide baseline and proposed approach.
        # table1_3
        python3 repair_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table1_3 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/DeepInspect_1/model_best.pth.tar
        python3 repair_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table1_3 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/table1_3/model_best.pth.tar --checkmodel
        # 0.9371; dog->cat:0.068;cat->dog: 0.06
        
        
        # table2_3
        python3 repair_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_3 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/DeepInspect_0/model_best.pth.tar
        python3 repair_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_3 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/table2_3/model_best.pth.tar --checkmodel
        # 0.9264; dog->cat:0.068;cat->dog: 0.056


        # table1_1
        python3 regular_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table1_1 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/DeepInspect_1/model_best.pth.tar
        python3 regular_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table1_1 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/table1_1/model_best.pth.tar --checkmodel
        # 0.9443; dog->cat:0.073;cat->dog: 0.075

        # table2_1
        python3 regular_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_1 --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ./runs/DeepInspect_0/model_best.pth.tar
        python3 regular_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_1 --epochs 60 --beta 1.0 --cutmix_prob 0 --pretrained ./runs/table2_1/model_best.pth.tar --checkmodel
        # 0.9346; dog->cat: 0.067; cat->dog: 0.06

        # table2_5
        python3 repair_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_5 --epochs 60 --beta 1.0 --cutmix_prob 0.0 --pretrained ./runs/DeepInspect_0/model_best.pth.tar
        python3 repair_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_5 --epochs 60 --beta 1.0 --cutmix_prob 0.0 --pretrained ./runs/table2_5/model_best.pth.tar --checkmodel
        # 0.9294; dog->cat: 0.057; cat->dog: 0.059

        # table1_4
        python3 target_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table1_4 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/DeepInspect_1/model_best.pth.tar
        python3 target_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table1_4 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/table1_4/model_best.pth.tar --checkmodel
        # 0.8398; 0.056; 0.082;
        # table2_4
        python3 target_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_4 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/DeepInspect_0/model_best.pth.tar
        python3 target_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_4 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/table2_4/model_best.pth.tar --checkmodel
        #0.8731; 0.068; 0.069;

        # table1_2
        python3 target_only_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table1_2 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/DeepInspect_1/model_best.pth.tar
        python3 target_only_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table1_2 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/table1_2/model_best.pth.tar --checkmodel
        # 0.8959; 0.075; 0.065

        # table2_2
        python3 target_only_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_2 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/DeepInspect_0/model_best.pth.tar
        python3 target_only_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname table2_2 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/table2_2/model_best.pth.tar --checkmodel
        # 0.9114; 0.055; 0.071


experiment2: 
    '''
    1. global seeds still cannot solve the deterministic issue.
        disabling cudnn make the training too slow.
    '''
    #step1: Train baseline models, with accuracy and confusion recorded in each epochs.
        python3 train_baseline.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname cifar10_resnet_2_1 --epochs 300 --beta 1.0 --cutmix_prob 1
        python3 repair_retrain_exp.py --net_type resnet --dataset cifar10 --depth 50 --batch_size 256 --lr 0.1 --expname cifar10_resnet_2_1 --epochs 60 --beta 1.0 --cutmix_prob 1.0 --pretrained ./runs/cifar10_resnet_2_1/model_best.pth.tar --checkmodel
        # accuracy: 0.943; dog->cat 0.064; cat->dog 0.053


run_script_cifar.py
import os
import time
from subprocess import PIPE, run

def execute(command):
    result = run(command, universal_newlines=True, shell=True, capture_output=True, text=True)
    # print(result.stdout)
    print(result.stderr)


method_properties = {
'w-aug': {
    "param_name": "weight",
    "filename": "exp_oversampling",
    "replace": "",
    "batch_size": 128,
    "extra_batch_size": 0,
},
'w-bn': {
    "param_name": "ratio",
    "filename": "exp_newbn",
    "replace": " --replace",
    "batch_size": 128,
    "extra_batch_size": 128,
},
'w-loss': {
    "param_name": "target_weight",
    "filename": "exp_weighted_loss",
    "replace": "",
    "batch_size": 128,
    "extra_batch_size": 10,
},
'w-dbr': {
    "param_name": "lam",
    "filename": "dbr",
    "replace": "",
    "batch_size": 128,
    "extra_batch_size": 10,
}
}



dataset_model_classes = [('cifar10', 'resnet-18', (3, 5, 2)), ('cifar10', 'vggbn-11', (3, 5, 2)), ('cifar100', 'resnet-34', (98, 35, 11))]
tasks = ['confusion', 'bias']
methods = ['w-aug', 'w-bn', 'w-loss', 'w-dbr']
params = [0.1, 0.3, 0.5, 0.7, 0.9]
epochs = 60
verbose = ""

dataset_model_classes = [('cifar10', 'resnet-18', (3, 5, 2))]
tasks = ['confusion']
methods = ['w-aug']
params = [0.1, 0.3, 0.5, 0.7, 0.9]
# epochs = 2


config_list = [
('cifar10', 'resnet-18', (3, 5, 2), 'confusion', 'w-aug', 0.7),
('cifar10', 'resnet-18', (3, 5, 2), 'confusion', 'w-bn', 0.9),
('cifar10', 'resnet-18', (3, 5, 2), 'confusion', 'w-loss', 0.9),
('cifar10', 'resnet-18', (3, 5, 2), 'confusion', 'w-dbr', 0.9),
('cifar10', 'vggbn-11', (3, 5, 2), 'confusion', 'w-aug', 0.9),
('cifar10', 'vggbn-11', (3, 5, 2), 'confusion', 'w-bn', 0.9),
('cifar10', 'vggbn-11', (3, 5, 2), 'confusion', 'w-loss', 0.9),
('cifar10', 'vggbn-11', (3, 5, 2), 'confusion', 'w-dbr', 0.7),
('cifar100', 'resnet-34', (98, 35, 11), 'confusion', 'w-aug', 0.7),
('cifar100', 'resnet-34', (98, 35, 11), 'confusion', 'w-bn', 0.9),
('cifar100', 'resnet-34', (98, 35, 11), 'confusion', 'w-loss', 0.7),
('cifar100', 'resnet-34', (98, 35, 11), 'confusion', 'w-dbr', 0.9),

('cifar10', 'resnet-18', (3, 5, 2), 'bias', 'w-aug', 0.7),
('cifar10', 'resnet-18', (3, 5, 2), 'bias', 'w-bn', 0.9),
('cifar10', 'resnet-18', (3, 5, 2), 'bias', 'w-loss', 0.9),
('cifar10', 'resnet-18', (3, 5, 2), 'bias', 'w-dbr', 0.9),
('cifar10', 'vggbn-11', (3, 5, 2), 'bias', 'w-aug', 0.9),
('cifar10', 'vggbn-11', (3, 5, 2), 'bias', 'w-bn', 0.9),
('cifar10', 'vggbn-11', (3, 5, 2), 'bias', 'w-loss', 0.7),
('cifar10', 'vggbn-11', (3, 5, 2), 'bias', 'w-dbr', 0.9),
('cifar100', 'resnet-34', (98, 35, 11), 'bias', 'w-aug', 0.7),
('cifar100', 'resnet-34', (98, 35, 11), 'bias', 'w-bn', 0.9),
('cifar100', 'resnet-34', (98, 35, 11), 'bias', 'w-loss', 0.9),
('cifar100', 'resnet-34', (98, 35, 11), 'bias', 'w-dbr', 0.9),
]



def execute_cmd(dataset, model, classes, task, method, param, log_filename, t0, rep_num):
    if model == 'vggbn-11':
        model_type = 'resnet'
        model_depth = '18'
        vggbn = '_vggbn'
    else:
        model_type, model_depth = model.split('-')
        vggbn = ''

    if model == 'vggbn-11':
        model_path = 'models/cifar10_vggbn_2_4/model_best.pth.tar'
    elif model == 'resnet-18':
        model_path = 'models/cifar10_resnet18_2_4/model_best.pth.tar'
    elif model == 'resnet-34':
        model_path = 'models/cifar100_resnet34/model_best.pth.tar'
    else:
        raise
    first, second, third = classes

    param_name = method_properties[method]["param_name"]
    filename = method_properties[method]["filename"]
    replace = method_properties[method]["replace"]
    batch_size = method_properties[method]["batch_size"]
    extra_batch_size = method_properties[method]["extra_batch_size"]

    filepath = os.path.join('exp_7', dataset, task, 'repair_'+task+'_'+filename+vggbn+'.py')
    expname = dataset+'_'+model+'_'+task+'_'+method+'_'+str(param)

    cmd = f"python3 {filepath} --net_type {model_type} --dataset {dataset} --depth {model_depth} --expname {expname} --epochs {epochs} --lr 0.1 --beta 1.0 --cutmix_prob 0 --pretrained {model_path} --batch_size {batch_size} --extra {extra_batch_size} --first {first} --second {second} --{param_name} {param}"+replace+verbose

    if task == 'bias':
        cmd += ' --third '+str(third)
    print('-'*20)
    print(cmd)
    print('-'*20)
    with open('tmp_log.txt', 'a') as f_out:
        f_out.write(cmd+'\n')
        f_out.write(str(time.time()-t0)+'\n')
    execute(cmd)

if __name__ == '__main__':
    # ['grid', 'specific']
    mode = 'specific'
    t0 = time.time()
    if mode == 'grid':
        log_filename = 'tmp_log.txt'
        with open(log_filename, 'w') as f_out:
            pass
        for dataset, model, classes in dataset_model_classes:
            for task in tasks:
                for method in methods:
                    for param in params:
                        execute_cmd(dataset, model, classes, task, method, param, log_filename, t0)
    elif mode == 'specific':
        rep_nums = 4
        log_filename = 'tmp_log_coco_specific.txt'
        with open(log_filename, 'w') as f_out:
            pass
        for rep_num in range(rep_nums):
            for dataset, model, classes, task, method, param in config_list:
                execute_cmd(dataset, model, classes, task, method, param, log_filename, t0, rep_num)
    else:
        raise
